---
title: 'Deconvolution of multi-region WGS colorectal samples'
author: "Giulio Caravagna"
date: "12/03/2020"
output:
  html_document:
    theme: cosmo
---


```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

We describe the analyses of the multi-region colorectal adenocarcinoma `Set7` (4 biopsies) and `Set6` (6 biopsies), for which we have generated new WGS sequencing data (~100x median coverage). The data - *purity, single nucleotide variants, copy number calls and clustering results* - are available in [Supplementary Table 2 (Excel)](./Supplementary_Table_S2.xlsx).

The data for these tumours has been first sequenced, at lower coverage, in [Cross W, et al. The evolutionary landscape of colorectal tumorigenesis. Nat Ecol Evol. 2018;2(10):1661â€“1672.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6152905/)

To implement this analysis we use the following packages:

* [MOBSTER](https://caravagn.github.io/mobster), to detect tails from multi-region data;
* [VIBER](https://caravagn.github.io/VIBER), to use a multivariate Binomial mixture model on read counts;
* [CNAqc](https://caravagn.github.io/CNAqc), integrate mutation and copy number data.

Please refer to the webpage of each one of them for installation instructions. 

# Analysis of Set7 (4 biopsies)

We discuss the analysis for `Set7`, one of the two colorectal samples. The output RDS objects generated by this vignette are:

* [Set7_mobster_fits.rds](./Set7/Set7_mobster_fits.rds), the fits per sample with MOBSTER;
* [Set7_mobster_viber_fit.rds](./Set7/Set7_mobster_viber_fit.rds), the VIBER fit, after MOBSTER;
* [Set7_mobster_viber_fit_heuristics.rds](./Set7/Set7_mobster_viber_fit_heuristics.rds), the VIBER fit, after MOBSTER and the cluster-filtering heuristics;
* [Set7_standard_fit.rds](./Set7/Set7_standard_fit.rds), the VIBER fit without MOBSTER;
* [Set7_standard_fit_heuristics.rds](./Set7/Set7_standard_fit_heuristics.rds), the VIBER fit without MOBSTER;

You need some auxiliary functions: [auxiliary_functions.R](./auxiliary_functions.R)

```{r, message=FALSE, warning=FALSE}
# Packages that are required specifically for the analysis
require(CNAqc)
require(mobster)
require(VIBER)
require(dplyr)

# Source a bunch of auxiliary functions
source('auxiliary_functions.R', verbose = FALSE)
```

# Loading the data

We begin loading the data from the CSV files `"Set7_mutations.csv"` and `"Set7_cna.csv"`, which are 
released along with the paper.

```{r, fig.width=8, fig.height=4, warning=FALSE}
dir.create("./Set7/")

###########################v
# Load data from CSV files #
############################

# Sample names and purity (from previous analysis we know purity)
Set7_samples = paste0('Set7_', c(55, 57, 59, 62))
Set7_purity = pio:::nmfy(Set7_samples, c(.88, .88, .88, .8))

# We load SNV data for diploid regions that map to CNA segmentes with >500 mutations. 
# These have been precomputed with the functions of the CNAqc package
Set7_mutations = readr::read_csv("./data/Set7_mutations.csv", col_types = readr::cols())

print(Set7_mutations)

# Load CNA data for all segments (not just diploid)
Set7_CNA = readr::read_csv("./data/Set7_cna.csv", col_types = readr::cols())

print(Set7_CNA)
```

We use the CNAqc package to visualise and visualise the CNA segments and the VAF distribution.

```{r, fig.height=6, fig.width=10, warning=F, message=T}
plot_calls(
  samples = Set7_samples, 
  CNA_calls = Set7_CNA, 
  mutation_calls = Set7_mutations, 
  purities = Set7_purity
  )
```

# Step one, removing neutral mutations

Fit each sample with MOBSTER using the raw VAF data, save the RDS output linked to this vignette, and plot the fits as (2x2 matrix).
```{r, fig.height=6, fig.width=6, warning=T, message=T, echo=T}
mobster_fits = fit_mobsters(Set7_mutations, Set7_samples)

# Save RDS
saveRDS(mobster_fits, file = "./Set7/Set7_mobster_fits.rds")

# Plot a 2x2 figure
ggarrange(plotlist = lapply(mobster_fits, plot), ncol = 2, nrow = 2)
```

We extract non-tail mutations, i.e., mutations never assigned to a tail.
```{r}
# From the MOBSTER fits
non_tail_mutations = get_nontail_mutations(mutations = Set7_mutations, mobster_fits)

print(non_tail_mutations)
```

We create separate inputs for the mutation read depth (DP, *coverage or total number of reads*)
and the number of reads with the alternative allele (NV, *number of variant reads*).

```{r}
# Cluster non-tail mutations with VIBER
DPs = non_tail_mutations %>% select(ends_with('DP'))
NVs = non_tail_mutations %>% select(ends_with('NV'))
  
colnames(DPs) = colnames(NVs) = Set7_samples 

# Coverage data
print(DPs)
```

# Cluster remaining read counts

We can now run the variational inference fitting methods implemented
in VIBER, to capture mixtures of Binomial distributions from non-tail SNVs.
```{r, fig.height=6, fig.width=6, warning=F, message=F, echo=T}
options(easypar.parallel = FALSE)

# VIBER fit, and save RDS
viber_fit = VIBER::variational_fit(x = NVs, y = DPs, samples = 2, epsilon_conv = 1e-6)

saveRDS(viber_fit, file = "./Set7/Set7_mobster_viber_fit.rds")

# Plot a 3x2 figure -- raw fit (all clusters)
ggarrange(
  plotlist = plot(
    viber_fit, 
    colors = get_cluster_colors(palettes = 'FantasticFox1', viber_fit)
    ), 
  ncol = 3, 
  nrow = 2)
```

As we discuss in the paper, we use a heuristic (implemented in VIBER)
to filter out some clusters, and re-perform the plots.
```{r, fig.height=6, fig.width=6, warning=F, message=F, echo=T}
# Apply the heuristic, and save another RDS
heuristic_fit = VIBER::choose_clusters(viber_fit, dimensions_cutoff = 1)

saveRDS(heuristic_fit, file = "./Set7/Set7_mobster_viber_fit_heuristics.rds")

# Plot a 3x2 figure -- after the heuristic
ggarrange(
  plotlist = plot(
    heuristic_fit, 
    colors = get_cluster_colors(palettes = 'FantasticFox1', heuristic_fit)
    ), 
  ncol = 3, 
  nrow = 2)
```

# Analysis without MOBSTER

The standard analysis without MOBSTER clusters read counts for all SNVs.

```{r, fig.height=6, fig.width=6, warning=F, message=F, echo=T}
# All mutations with VIBER
DPs = Set7_mutations %>% select(ends_with('DP'))
NVs = Set7_mutations %>% select(ends_with('NV'))

colnames(DPs) = colnames(NVs) = Set7_samples 

# VIBER fit, and save RDS
st_viber_fit = VIBER::variational_fit(x = NVs, y = DPs, samples = 1, epsilon_conv = 1e-6, max_iter = 100)

saveRDS(st_viber_fit, file = "./Set7/Set7_standard_fit.rds")

# Plot a 3x2 figure -- before the heuristic
ggarrange(
  plotlist = plot(
    st_viber_fit, 
    colors = get_cluster_colors(palettes = c('BottleRocket2', "Zissou1"), st_viber_fit)
    ), 
  ncol = 3, 
  nrow = 2)

# Apply the heuristic
st_heuristic_fit = VIBER::choose_clusters(st_viber_fit, dimensions_cutoff = 1)

saveRDS(st_heuristic_fit, file = "./Set7/Set7_standard_fit_heuristics.rds")

# Plot a 3x2 figure -- after the heuristic
ggarrange(
  plotlist = plot(
    st_heuristic_fit, 
    colors = get_cluster_colors(palettes = c('BottleRocket2', "Zissou1"), st_heuristic_fit)
    ), 
  ncol = 3, 
  nrow = 2)
```

# Summary comparive fit plots 

We use an auxiliary plot assembly function that places on the top and bottom diagonal
two VIBER fits, and MOBSTER fits on the diagonal of the matrix.


Without versus with MOBSTER, without the heuristic.

```{r, fig.height=16, fig.width=15, warning=F, message=F, echo=T}
squareplot(
  mobster_fits = mobster_fits, 
  viber_fit_bottom = st_viber_fit, 
  viber_fit_top = viber_fit, 
  samples_list = Set7_samples,
  colors_bottom = get_cluster_colors(palettes = c('BottleRocket2', "Zissou1"), st_viber_fit),
  colors_top = get_cluster_colors(palettes = 'FantasticFox1', viber_fit)
  )
```

Without versus with MOBSTER, with the heuristic.

```{r, fig.height=16, fig.width=15, warning=F, message=F, echo=T}
squareplot(
  mobster_fits = mobster_fits,
  viber_fit_bottom = st_heuristic_fit,
  viber_fit_top = heuristic_fit,
  samples_list = Set7_samples,
  colors_bottom = get_cluster_colors(palettes = c('BottleRocket2', "Zissou1"), st_heuristic_fit),
  colors_top = get_cluster_colors(palettes = 'FantasticFox1', heuristic_fit)
)
```

# Analysis of Set6 

The output RDS objects generated by this vignette are:

* [Set6_mobster_fits.rds](./Set6/Set6_mobster_fits.rds), the fits per sample with MOBSTER;
* [Set6_mobster_viber_fit.rds](./Set6/Set6_mobster_viber_fit.rds), the VIBER fit, after MOBSTER;
* [Set6_mobster_viber_fit_heuristics.rds](./Set6/Set6_mobster_viber_fit_heuristics.rds), the VIBER fit, after MOBSTER and the cluster-filtering heuristics;
* [Set6_standard_fit.rds](./Set6/Set6_standard_fit.rds), the VIBER fit without MOBSTER;
* [Set6_standard_fit_heuristics.rds](./Set6/Set6_standard_fit_heuristics.rds), the VIBER fit without MOBSTER;

```{r, echo=T, warning=FALSE, message=FALSE}
dir.create("./Set6/")

# Load sample names and purity (from previous analysis, we know purity)
Set6_samples = paste0('Set6_', c(42, 44, 45:48))
Set6_purity = pio:::nmfy(Set6_samples, c(0.66, 0.72, 0.80, 0.80, 0.80, 0.80))

# Load SNV and CNA data, as for Set7
Set6_mutations = read_csv('./data/Set6_mutations.csv')
Set6_CNA = read_csv('./data/Set6_cna.csv')

# MOBSTER fits
mobster_fits = fit_mobsters(Set6_mutations, Set6_samples)
saveRDS(mobster_fits, file = "./Set6/Set6_mobster_fits.rds")

# Non tails
non_tail_mutations = get_nontail_mutations(mutations = Set6_mutations, mobster_fits)

# VIBER, plus heuristic
DPs = non_tail_mutations %>% select(ends_with('DP'))
NVs = non_tail_mutations %>% select(ends_with('NV'))
colnames(DPs) = colnames(NVs) = Set6_samples 

viber_fit = VIBER::variational_fit(x = NVs, y = DPs, samples = 2, epsilon_conv = 1e-6)
saveRDS(viber_fit, file = "./Set6/Set6_mobster_viber_fit.rds")

heuristic_fit = VIBER::choose_clusters(viber_fit, dimensions_cutoff = 1)
saveRDS(heuristic_fit, file = "./Set6/Set6_mobster_viber_fit_heuristics.rds")

# Analysis without MOBSTER, plus heuristic
DPs = Set6_mutations %>% select(ends_with('DP'))
NVs = Set6_mutations %>% select(ends_with('NV'))
colnames(DPs) = colnames(NVs) = Set6_samples 

# VIBER fit
st_viber_fit = VIBER::variational_fit(x = NVs, y = DPs, samples = 1, epsilon_conv = 1e-6, max_iter = 100)
saveRDS(st_viber_fit, file = "./Set6/Set6_standard_fit.rds")

st_heuristic_fit = VIBER::choose_clusters(st_viber_fit, dimensions_cutoff = 1)
saveRDS(st_heuristic_fit, file = "./Set6/Set6_standard_fit_heuristics.rds")
```

Without versus with MOBSTER, without the heuristic.

```{r, fig.height=20, fig.width=17, warning=F, message=F, echo=T}
squareplot(
  mobster_fits = mobster_fits, 
  viber_fit_bottom = st_viber_fit, 
  viber_fit_top = viber_fit, 
  samples_list = Set6_samples,
  colors_bottom = get_cluster_colors(palettes = c('BottleRocket2', "Zissou1"), st_viber_fit),
  colors_top = get_cluster_colors(palettes = 'FantasticFox1', viber_fit)
  )
```

Without versus with MOBSTER, with the heuristic.

```{r, fig.height=20, fig.width=17, warning=F, message=F, echo=T}
squareplot(
  mobster_fits = mobster_fits, 
  viber_fit_bottom = st_heuristic_fit, 
  viber_fit_top = heuristic_fit, 
  samples_list = Set6_samples,
  colors_bottom = get_cluster_colors(palettes = c('BottleRocket2', "Zissou1"), st_heuristic_fit),
  colors_top = get_cluster_colors(palettes = 'FantasticFox1', heuristic_fit)
  )
```

